### ISSUE TITLE
Feature Request: Support Gitlab repositories

### ISSUE THREAD

**Issue by krisztiankocsis** [getsentry/sentry#5622]
Hi!

The new repository integration for release commits seems interesting to us. Unfortunately currently generic Git repositories or Gitlab repositories not supported. Pls. add support for Gitlab repositories or generic remote Git repositories (e.g. via HTTPS with user/password authentication at least).

**Comment by dcramer** [comment]
It's likely we'll be able to support GitLab, but we're first working on improving GitHub, and then landing BitBucket. At some point we intend to support generic repositories as well, however due to the complexity and security constraints required to host large volumes of repositories we're avoiding that problem right now.

**Comment by brodock** [comment]
@dcramer Anything we can help from GitLab's side?

**Comment by arvalaan** [comment]
+1

**Comment by sw-double** [comment]
This quickly became the most upvoted issue, it probably deserves some priority and planning.

**Comment by jmagnusson** [comment]
I would so love to have this. My too favorite code services deserve a good relationship 😊

**Comment by CowChris** [comment]
+1 Please integrate this. Please please please!

**Comment by dcramer** [comment]
This is something we'd like to get done yet this year, but I'll be open and say its not high priority. A lot of what we're doing is deepening what the data around VCS integration gets us yet, and given we have a lot less GitLab users than e.g. GitHub/Bitbucket, we're working on polishing those up before expanding implementations.

Aside if someone from the community is interested in taking this up, it's not overly complex and we'd be happy to provide guidance. The complexity is that a lot of GitLab lives on premise and we can't support those installations until some new work is done (e.g. #5963).

**Comment by lig** [comment]
@dcramer how do you count Gitlab vs Github users?

**Comment by dcramer** [comment]
@lig we look at two things:

1. GitLab's goal is fundamentally On-Premise installations, and most of the rest of their business is free users. Since we dont make money on either of those two, it by default is a lower priority.
2. We look at usage statistics and customer requests. GitLab comes up every so often, but we have a lot of users of GitHub for e.g. the Issue Tracking integration.

We're currently rewriting the integrations APIs to make it quicker and easier to support things going forward, but I will say I was able to get a Visual Studio Team Services repository provider ready pretty quickly, so it's likely we could do similar with GitLab if the APIs are all there.

**Comment by rvanlaak** [comment]
Although the arguments aren't invalid, they don't make sense from Sentry's perspective. Users that have the free Gitlab version might be absolutely fine with paying for Senty, as that's a complete different service.

Personally I would see integrations as a huge marketing machine for choosing Sentry, as happened with Slack and it's integrations ecosystem. So, great to hear about the integrations API rewrite. Keep up the good work!

Maybe the first step should be to make a quick analysis of their API, as I know that is really extensive, and probably - except for it's most recently released features - exposes all Gitlab's data: https://docs.gitlab.com/ee/api/README.html

**Comment by lig** [comment]
@rvanlaak I agree. As a small company, we have free Gitlab, self-hosted Gitlab CI and self-hosted Sentry. However, we pay for Taiga.io for instance, just because its handy to use it as a service. I see it is fairly acceptable for us to migrate to Sentry as a service and Gitlab EE if this saves us support time.
Nevertheless, @dcramer, keep up the good work!

**Comment by rakshazi** [comment]
Hi guys,
any update on this?

**Comment by EvaSDK** [comment]
Here is a small script I used at my company, maybe someone with a bit more insight on Sentry codebase can turn it into a plugin.

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import requests

GITLAB_URL = 'https://gitlab.company.tld'
GITLAB_API_URL = GITLAB_URL + '/api/v4'
GITLAB_API_TOKEN = 'gitlabtoken'
GITLAB_PROJECT_NAME = 'gitlabgroup/project'
GITLAB_PROJECT_NAME_ENC = GITLAB_PROJECT_NAME.replace('/', '%2F')
SENTRY_API_URL = 'https://sentry.company.tld/api/0'
SENTRY_API_TOKEN = 'sentrytoken'
SENTRY_PROJECT = 'dummy-proj-slug'
SENTRY_ORGANIZATION = 'dummy-org'

tags = []
page = 1
while page:
    res = requests.get(
        GITLAB_API_URL + '/projects/' + GITLAB_PROJECT_NAME_ENC + '/repository/tags',
        params={
            'page': page,
            'per_page': 100,
        },
        headers={
            'Private-Token': GITLAB_API_TOKEN,
        },
    )
    res.raise_for_status()
    tags.extend(res.json())
    page = res.headers.get('X-Next-Page', 0)

tags = list(sorted(tags, key=lambda tag: tag['commit']['created_at']))
tags = tags[-4:]
print('Sorted releases:', tags)
print('>>> Found a total of', len(tags), 'tags')

previous_tag, tags = tags[0], tags[1:]
for tag in tags:
    print('>>> Reading changes from', previous_tag['name'], 'to', tag['name'])
    res = requests.get(
        GITLAB_API_URL + '/projects/' + GITLAB_PROJECT_NAME_ENC + '/repository/compare',
        params={
            'from': previous_tag['name'],
            'to': tag['name'],
        },
        headers={
            'Private-Token': GITLAB_API_TOKEN,
        },
    )
    commits = res.json()['commits']
    res.raise_for_status()
    if not commits:
        print('>>> No commits to submit')
        continue

    print('>>> Submitting changes from', previous_tag['name'], 'to', tag['name'])
    data = {
        'version': tag['name'],
        'ref': tag['commit']['id'],
        'url': GITLAB_URL + '/' + GITLAB_PROJECT_NAME + '/tags/' + tag['name'],
        'projects': [SENTRY_PROJECT],
        'dateReleased': tag['commit']['created_at'],
        'commits': [{
            'id': commit['id'],
            'repository': GITLAB_PROJECT_NAME,
            'message': commit['title'].strip(),
            'author_name': commit['author_name'],
            'author_email': commit['author_email'],
            'timestamp': commit['created_at'],
        } for commit in commits],
    }
    res = requests.post(
        SENTRY_API_URL + '/organizations/' + SENTRY_ORGANIZATION + '/releases/',
        json=data,
        headers={'Authorization': 'Bearer {}'.format(SENTRY_API_TOKEN)},
    )
    res.raise_for_status()
    previous_tag = tag
```

**Comment by andrzejwp** [comment]
@EvaSDK could you please give us a little background on how you integrated this with your sentry?

**Comment by rakshazi** [comment]
@EvaSDK good idea :) we use @titanium-codes/cibu (command: sentry release) with @tisuit/skeleton .gitlab-ci.yml
@andrzejwp EvaSDK's script loads git tags from gitlab repo via API and creates sentry releases based on these tags

**Comment by Jean85** [comment]
> We're currently rewriting the integrations APIs to make it quicker and easier to support things going forward, but I will say I was able to get a Visual Studio Team Services repository provider ready pretty quickly, so it's likely we could do similar with GitLab if the APIs are all there.

@dcramer if you can tell us what info Sentry needs to scrape, we can review GitLab's API for you and create a list of the needed endpoints. From what I know the integrations should be straightforward, the API are pretty rich and OAuth2 is well supported (and already used by Sentry elsewhere too)

**Comment by dcramer** [comment]
@Jean85 realistically we'd just need the same thing that we implemented for the GitHubRepositoryProvider (in sentry-plugins).

1. [recommended] webhook (per push, but at least per commit)
2. endpoint to query a range of commits (e.g. ``HEAD..abcdef``)
3. endpoint to query a set of commits (or at least an individual commit)
4. OAuth with permissions that don't give access to write to a repository

**Comment by brodock** [comment]
@dcramer Here is the documentation for each of your itens:

1- https://docs.gitlab.com/ce/user/project/integrations/webhooks.html#push-events
2- https://docs.gitlab.com/ee/api/repositories.html#compare-branches-tags-or-commits
3- https://docs.gitlab.com/ee/api/commits.html#get-a-single-commit (if you need multiple commits in a single request, please let us know in a feature request issue)
4- https://docs.gitlab.com/ee/integration/oauth_provider.html#authorized-applications we don't have a 'read_only' type of scope yet, it should work fine with "api" scope

You can configure the webhook per project entering it manually or using a service template, which can also be configured to be include automatically in each new project on per GitLab instance level by a System Administrator. Read more here: https://docs.gitlab.com/ee/user/project/integrations/project_services.html

**Comment by iSDP** [comment]
Really hoping for this to be added.

Seems doable to say the least. 

+1

**Comment by llonchj** [comment]
+1

**Comment by fprochazka** [comment]
Guys, spaming +1 will only make the devs lock this thread (since ~600 people are watching it). Either be constructive with suggestions or use the reaction emojis. Thanks!

**Comment by metaa** [comment]
@dcramer can we try to add this into the 9.0 milestone?
Would be awesome if this issue won't be lost forgotten.

**Comment by dcramer** [comment]
It won’t happen for our 9 release but we’re hoping to make it easy as an immediate follow up. Unfortunately I’m still not sure we can commit to building it ourselves as there are a lot of integrations and we have to prioritize the most impactful.

**Comment by JohnMH** [comment]
@dcramer While I understand that it's hard to get good numbers, I think it's safe to say that even if an eighth of the people that use GitHub use GitLab (either GitLab.com, easier to get numbers for, or self-hosted), it should be embraced just as GitHub.

If you could point people in the right direction, I am certain that members of the community would be happy to work on this feature.

**Comment by dcramer** [comment]
@JohnMH it comes down to "what can we do, and whats most important". GitHub, BitBucket, and VSTS are our first priority and they all have to be updated to work best with our upcoming Sentry 9 features. Even beyond that, most GitLab users (at least anyone that would pay for Sentry) run it on-premise. We'll explore providing a guide to integrations after we get everything out the door, but its still our intention to support the 80% use cases as first class citizens in Sentry. Right now we're really just focused on getting scaffolding in place, and not adding additional third parties that we have to maintain integrations for.

**Comment by iSDP** [comment]
I'm using GitLab extensively.

I'm really hoping for proper intergration to be added, if only the commit tracking..

**Comment by metaa** [comment]
I just set up sentry for new projects for my company a few weeks back and everything we do is in GitLab.

Commit tracking is basically the only thing we can't use because of this 😢

**Comment by stanhu** [comment]
I quickly hacked a proof of concept to show that it's not all that difficult to get this to work with GitLab in my work-in-progress branch in https://github.com/getsentry/sentry-plugins/pull/319. It doesn't support Webhook processing at the moment, but it does retrieve commits like so:

![image](https://user-images.githubusercontent.com/963826/36085256-663b7f26-0f79-11e8-9e10-cf523c0cbec2.png)

Things to do:

* [ ] Support GitLab as an authorization provider
* [ ] Support instance-specific URLs

**Comment by dcramer** [comment]
@stanhu the main thing is you cant reasonably support instance specific URLs. We're intending to remove the social_auth (current gen bindings) as we have a new solution coming. Its the primary blocker for GitLab / GHE.

**Comment by RomanTsegelskyi** [comment]
@dcramer was this scheduled?

**Comment by JohnMH** [comment]
@stanhu I haven't take a look yet (haven't had the time), but would your code work if it got the GitLab instance URL from a configuration file? Perhaps from Sentry's config? I am certain that 99% of on-premise users would be able to use that, as most that would also be running Sentry on-premise would only be using the one GitLab instance.

**Comment by dcramer** [comment]
If and when GitLab support comes to Sentry core it won't be until later in the second quarter of this year.

We won't be accepting any pull requests that don't fully implement the API in a future proof way.  That means we won't allow any changes to Sentry which *only* support Sentry if you're running it On-Premise, and we won't allow any changes which don't support at least gitlab.com (in a future proof way), and ideally also GitLab on-premise installations. Because the latter is very hard in v1/v2 of the plugin architecture, I wouldn't expect anyone outside of Sentry core to build this. At least not until our new integrations framework has been stabilized which makes all of this possible.

**Comment by asimon-startnetco** [comment]
Are there any news on this subject? We use Gitlab with Gitlab CI for most of our projects now and integration with Sentry is the only thing we're missing now.

**Comment by Majkl578** [comment]
> If and when GitLab support comes to Sentry core it won't be until later in the second quarter of this year.

It's already _later in the second quater_, any update?
Our company migrated from BitBucket to GitLab so this would really be helpful. :blush:

**Comment by andrewtch** [comment]
Any news on that?

**Comment by fones** [comment]
I am thinking about setting up bitbucket and push code to gitlab and bitbucket in the same time. What features I will miss doing this way?

**Comment by Coffee2CodeNL** [comment]
I'm waiting for the new plugin API before I continue working on this.

**Comment by dcramer** [comment]
@Coffee2CodeNL fwiw we'll probably jsut build it in house at this point (tbd yet). Right now we're actively working on GitHub, BitBucket, Jira, and VSTS, and we plan to follow up with GitLab immediately after. It's hopefully going to be very quick since its not all that different from GitHub.

**Comment by marcelloceschia** [comment]
If someone need the GitLab repository support for 8.22 on premise, I have modefied the version provided by stanhu to support dynamic configuration.
So no need for static url and token hacks within the plugin file.

For Docker deployment you can add two lines to your Dockerfile

```
ADD gitlab/client.py /usr/local/lib/python2.7/site-packages/sentry_plugins/gitlab/
ADD gitlab/plugin.py /usr/local/lib/python2.7/site-packages/sentry_plugins/gitlab/
```

and you are ready

**Comment by marcelloceschia** [comment]
I've added some improvements and fixes to the merge request. The patch_set will be added to the commit as well. Can you check this integration please.

**Comment by muratcorlu** [comment]
With the latest news and movement, I hope this task will be more prioritized. https://twitter.com/hashtag/movingtogitlab

**Comment by dcramer** [comment]
We will update this ticket when GitLab supportnis available, but until then I’m locking it as the comments have distracting.

**Comment by markstory** [comment]
Closing this as the GitLab integration is in master and will be available to all SaaS customers very soon.
